src/cockpit/channels/filesystem.py

```diff
@@ -268,3 +287,226 @@ def do_close(self):
+                      stat.S_IFBLK: 'blk', stat.S_IFIFO: 'fifo', stat.S_IFSOCK: 'sock'}
+        available_stat_getters = {
+            'type': lambda buf: buf and stat_types.get(stat.S_IFMT(buf.st_mode)),
+            'tag': tag_from_stat,
+            'mode': lambda buf: stat.S_IMODE(buf.st_mode),
+            'size': lambda buf: buf.st_size,
+            'uid': lambda buf: buf.st_uid,
+            'gid': lambda buf: buf.st_gid,
+            'mtime': lambda buf: buf.st_mtime,
+            'user': lambda buf: get_user(buf.st_uid),
+            'group': lambda buf: get_group(buf.st_gid),
+        }
+        stat_getters = tuple((key, available_stat_getters.get(key, lambda _: None)) for key in attrs)
+
+        def get_attrs(fd: int, name: str) -> 'JsonDict | None':
+            try:
+                buf = os.lstat(name, dir_fd=fd) if name else os.fstat(fd)
+            except FileNotFoundError:
+                return None
+            except OSError:
+                return {name: None for name, func in stat_getters}
+
+            result = {key: func(buf) for key, func in stat_getters}
+
+            if 'target' in result and stat.S_IFMT(buf.st_mode) == stat.S_IFLNK:
+                with contextlib.suppress(OSError):
+                    result['target'] = os.readlink(name, dir_fd=fd)
+
+            return result
+
+        return get_attrs
+
+    def send_update(self, updates: JsonDict, *, reset: bool = False) -> None:
+        if reset:
+            if set(self.current_value) & set(updates):
+                # if we have an overlap, we need to do a proper reset
+                self.send_json({name: None for name in self.current_value}, partial=True)
+                self.current_value = {'partial': True}
+                updates.update(partial=None)
+            else:
+                # otherwise there's no overlap: we can just remove the old keys
+                updates.update({key: None for key in self.current_value})
+
+        json_merge_and_filter_patch(self.current_value, updates)
+        if updates:
+            self.send_json(updates)
+
+    def process_update(self, updates: 'set[str]', *, reset: bool = False) -> None:
+        assert self.fd is not None
+
+        entries: JsonDict = {name: self.getattrs(self.fd, name) for name in updates}
+        info = entries.pop('', {})
+        assert isinstance(info, dict)  # fstat() will never fail with FileNotFoundError
+
+        if self.effective_fnmatch:
+            info['entries'] = entries
+
+        self.send_update({'info': info}, reset=reset)
+
+    def process_pending_updates(self) -> None:
+        assert self.pending is not None
+        if self.pending:
+            self.process_update(self.pending)
+        self.pending = None
+
+    def interesting(self, name: str) -> bool:
+        if name == '':
+            return True
+        else:
+            # only report updates on entry filenames if we match them
+            return fnmatch.fnmatch(name, self.effective_fnmatch)
```

Comment:
We'll arrive here very often for directories, even if we didn't set an fnmatch.  systemd_types gives no control over the events that we monitor on the `path` that we watch, so even if we're only interested in monitoring a directory (and not its child items) we're going to get a lot of create/delete/etc. events.  This function is the very thing that filters them — and an empty fnmatch indeed matches nothing.

I had a special case here before to specifically check for empty string and return fast, but I decided the code looked nicer this way.  I didn't measure performance either way, but I don't imagine it would be a significant difference.

Reply:
OK, then the `and self.effective_fnmatch` would make this conceptually cleaner, and possibly also faster. I don't insist though, so feel free to just resolve if you want to keep it as-is. Thanks for checking!